app/src/main/java/dev/arkbuilders/rate/presentation/quick/glancewidget/action/NextPageAction.kt

```diff
@@ -0,0 +1,35 @@
+package dev.arkbuilders.rate.presentation.quick.glancewidget.action
+
+import android.content.Context
+import androidx.glance.GlanceId
+import androidx.glance.action.ActionParameters
+import androidx.glance.appwidget.action.ActionCallback
+import androidx.glance.appwidget.state.updateAppWidgetState
+import dev.arkbuilders.rate.di.DIManager
+import dev.arkbuilders.rate.presentation.quick.glancewidget.QuickPairsWidget
+import dev.arkbuilders.rate.presentation.quick.glancewidget.QuickPairsWidgetReceiver.Companion.currentGroupKey
+
+class NextPageAction : ActionCallback {
+    override suspend fun onAction(
+        context: Context,
+        glanceId: GlanceId,
+        parameters: ActionParameters,
+    ) {
+        val quickRepo = DIManager.component.quickRepo()
+        val allGroups = quickRepo.getAllGroups()
+        updateAppWidgetState(context, glanceId) { prefs ->
+            val currentIndex = allGroups.indexOf(prefs[currentGroupKey])
+            var newIndex = if (currentIndex == -1) 0 else currentIndex + 1
+            if (newIndex > allGroups.lastIndex) {
+                newIndex = 0
+            }
+            val newGroup = allGroups[newIndex]
+            newGroup?.let {
+                prefs[currentGroupKey] = newGroup
+            } ?: let {
+                prefs.remove(currentGroupKey)
+            }
+        }
+        QuickPairsWidget().update(context, glanceId)
```

Comment:
With this particular code block, for `NextPageAction` and `PreviousPageAction`, there's only difference in the way we find `newIndex`. Hence we could share the rest of the code and only . HOC is a good fit in this case.

So we can have a helper file with this function:
```kotlin

suspend fun updatePinnedPairsWidgetNewPageState(
    context: Context,
    glanceId: GlanceId,
    quickRepo: QuickRepo,
    findNextIndexAction: (Int, Int) -> Int,
) {
    val allGroups = quickRepo.getAllGroups()
    updateAppWidgetState(context, glanceId) { prefs ->
        val currentIndex = allGroups.indexOf(prefs[QuickPairsWidgetReceiver.currentGroupKey])
        val newIndex = findNextIndexAction(currentIndex, allGroups.lastIndex)
        val newGroup = allGroups[newIndex]
        if (newGroup != null) {
            prefs[QuickPairsWidgetReceiver.currentGroupKey] = newGroup
        } else {
            prefs.remove(QuickPairsWidgetReceiver.currentGroupKey)
        }
    }
    QuickPairsWidget().update(context, glanceId)
}
```

And then update code in our two classes for example:
```kotlin
class NextPageAction : ActionCallback {
    override suspend fun onAction(
        context: Context,
        glanceId: GlanceId,
        parameters: ActionParameters,
    ) {
        val quickRepo = DIManager.component.quickRepo()
        updatePinnedPairsWidgetNewPageState(
            context = context,
            glanceId = glanceId,
            quickRepo = quickRepo,
            findNextIndexAction = { current, last ->
                findIndex(current, last)
            },
        )
    }

    private fun findIndex(
        currentIndex: Int,
        lastIndex: Int,
    ): Int {
        var newIndex = if (currentIndex == -1) 0 else currentIndex + 1
        if (newIndex > lastIndex) {
            newIndex = 0
        }
        return newIndex
    }
}
```